runtime/compiler/z/codegen/S390PrivateLinkage.cpp

```diff
@@ -1975,41 +1975,34 @@ J9::Z::PrivateLinkage::buildVirtualDispatch(TR::Node * callNode, TR::RegisterDep
          }
       else
          {
-         cursor =
-            generateRXInstruction(cg(), TR::InstOpCode::getExtendedLoadOpCode(), callNode, RegRA,
-                                   generateS390MemoryReference(classReg, offset, cg()));
+         // It should be impossible to have a offset that can't fit in 20bit given Java method table limitations.
+         // We assert here to insure limitation/assumption remains true. If this fires we need to fix this code
+         // and the _virtualUnresolvedHelper() code to deal with a new worst case scenario for patching.
+         TR_ASSERT_FATAL(offset>MINLONGDISP, "JIT VFT offset does not fit in 20bits");
+         TR_ASSERT_FATAL(offset!=0 || unresolvedSnippet, "Offset is 0 yet unresolvedSnippet is NULL");
+         TR_ASSERT_FATAL(offset<=MAX_IMMEDIATE_VAL, "Offset is larger then MAX_IMMEDIATE_VAL");
+
+         // If unresolved/AOT, this instruction will be patched by _virtualUnresolvedHelper() with the correct offset
+         cursor = generateRXInstruction(cg(), TR::InstOpCode::getExtendedLoadOpCode(), callNode, RegRA,
+                                        generateS390MemoryReference(classReg, offset, cg()));
 
          if (unresolvedSnippet)
             {
             ((TR::S390VirtualUnresolvedSnippet *)unresolvedSnippet)->setPatchVftInstruction(cursor);
             }
 
-         /* A load immediate (LHI/LGHI) instruction MUST be generated here for all virtual calls because
-          *
-          * 1. picbuilder looks for this load immediate (and an LG before it) and patches in the correct offset at
-          * runtime for unresolved virtual calls
-          *
-          * 2. for all virtual calls, resolved or not, this load immediate is used by the VM to find where
-          * the method is in the VFT table when doing J2I transitions. We need this negative offset in a register
-          * because S390 load address instruction used to find VFT table entires can't handle negative offsets.
-          * (VFT table precedes the J9Class in memory; as a result of this, all VFT entries are some negative offsets
-          * away from the J9Class)
-          *
-          *
-          * In fact we should not be using APIs (such as genLoadLongConstant) at all and be explicitly generating
-          * LHI/LGHI instruction. Similar occurrence below.
-          */
-         if (offset >= (-32768))
+         // A load immediate into R0 instruction (LHI/LGFI) MUST be generated here because the "LA" instruction used by
+         // the VM to find VFT table entries can't handle negative displacements. For unresolved/AOT targets we must assume
+         // the worse case (offset can't fit in 16bits). VFT offset 0 means unresolved/AOT, otherwise offset is negative.
+         // Some special cases have positive offsets i.e. java/lang/Object.newInstancePrototype()
+         if (!unresolvedSnippet && offset >= MIN_IMMEDIATE_VAL && offset <= MAX_IMMEDIATE_VAL) // Offset fits in 16bits
```

Comment:
I don't understand this comment. Can you elaborate what "the "LA" instruction used by the VM to find VFT table entries can't handle negative displacements" means?

Also could you explain why these special cases with positive VFT offsets exist? I have never seen this so I would like to understand how exactly this works.

Reply:
"LA instruction..." -- The VM uses an LA instruction in the J2I transition helper runtime code to add the VFT offset to the resolved class address in order to load the VFT entry that we want to execute. Since the LA instruction uses an unsigned immediate displacement the instruction is not able to add negative immediate numbers to the class address. Therefore we need to use a 0 for the immediate and r0 (containing the negative offset) as the index register in the LA instruction.

Special cases with positive offsets: For some calls where we have hard-coded helpers to implement the calle, we use positive offsets into the J9Class to find the JIT library method address we need to call. For these cases we'll have a positive offset from the class. For unresolved cases we'll have a 0 offset and for all others we'll have negative offsets.

Hope this helps to clarify things a bit.

Reply:
> "LA instruction..." -- The VM uses an LA instruction in the J2I transition helper runtime code to add the VFT offset to the resolved class address in order to load the VFT entry that we want to execute. Since the LA instruction uses an unsigned immediate displacement the instruction is not able to add negative immediate numbers to the class address. Therefore we need to use a 0 for the immediate and r0 (containing the negative offset) as the index register in the LA instruction.

I don't think the VM using an `LA` instruction matters here really. The J2I transition code is static and read only so the only option is to pass the VFT offset via a register. There is no alternative. We are effectively dispatching a vritual call that may branch to a JIT method or J2I transition. We don't know which one until runtime, so we have to handle both. That means we have to always load the VFT offset when performing the call because we may call the J2I thunk at runtime and in that case the VM needs to know which VFT entry to dispatch to, this is what the offset represents.

The comment should be altered IMO as whether the VM uses an `LA` instruction to perform the addition is not relevant. We can take a version of the previous paragraph as a summary of why the load is needed.

> Special cases with positive offsets: For some calls where we have hard-coded helpers to implement the calle, we use positive offsets into the J9Class to find the JIT library method address we need to call. For these cases we'll have a positive offset from the class. For unresolved cases we'll have a 0 offset and for all others we'll have negative offsets.

Thanks! Could we write this in as a comment in the code to explain when and why the VFT offset can be positive? That will help future readers who need to make changes in this area.

Reply:
Unless there is a unique thunk for each VFT offset then it does not make sense that the unsigned immediate in LA is the reason for setting up R0 in the JIT'ed code. I believe other platforms inspect the JIT'ed assembler code to get the offset, but even that technique would not limit using LA since it still can't be an immediate. The original comment before I shortened it has this reasoning.  

I'll add a comment about the possible values for offset.

Reply:
Sorry I'm still not following. What is this `LA` instruction you are referring to? I just debugged this with an example to convince myself of how it works. Here is the call flow:

1. We load the VFT offset into `GPR0`
2. We branch to the target method stored in the VFT offset memory location (some negative offset off the `J9Class*`) via the `BASR` instruction in the JIT mainline code
3. If the target method has not been compiled the pointer stored in the VFT offset will be the pointer of the `icallVMprJavaSendPatchupVirtual` function [1]. This function stores the value of `GPR0` into `J9VMThread.tempSlot` and then makes a call to `old_slow_icallVMprJavaSendPatchupVirtual` [2]
4. The `old_slow_icallVMprJavaSendPatchupVirtual` function takes the `J9VMThread.tempSlot` which is the VFT offset and the return address and calls `jitVTableIndex` which should return the VFT index. Notice how on z/Architecture this function just returns the second parameter [3]? We do not decode the instruction stream on Z, so it appears. This is not true on other platforms.
5. Execution returns back to `old_slow_icallVMprJavaSendPatchupVirtual` which then takes the VFT index and looks up the thunk based on the signature of the method we want to call. These thunks are one per signature and dispatch the calls from JIT to interpreter. The rest of this function call looks up the thunk for the particular call we want to make and then it overwrites the VFT entry in the `J9Class*` with the address of the thunk. This is so we don't have to pay the cost of looking up the thunk the next time we have to dispatch this virtual call. The last thing this method does is store the thunk value to `J9VMThread.tempSlot` [4].
6. Execution returns to `icallVMprJavaSendPatchupVirtual` which then dispatches a call to the thunk via an indirect branch off of `J9VMThread.tempSlot` [5]
7. The thunk in my case is just a simple instruction sequence which takes care of putting arguments in the right place on the stack, and then it calls `j2iVirtual` [6]. This ends up branching to the interpreter loop to execute the method.

As far as I can see the only reason why we are loading the VFT offset in a register is because looking at the implementation of `j2iVirtual` [6] the VM seems to want this VFT offset to be in `J9VMThread.tempSlot` [7]. It seems because of the possibility of having to dispatch via an interface. I don't quite understand the details of how this works, but I definitely would like to.

For me two questions arise:

1. Why is the implementation of `jitVTableIndex` more complicated on non-Z platforms? What is this API supposed to do exactly?
2. Why does the VM need the VFT offset in `J9VMThread.tempSlot`? Surely for resolved calls we should be able to decode this value from the JIT instruction stream. Given that JIT to interpreter calls are supposed to be rare, I'd like to know if there is an opportunity to remove the VFT load prior to virtual calls on all platforms as that can save quite a bit of pathlength.

I'm going to copy this comment into an issue so we can continue discussing there. However as for the review comment here I still don't see what this `LA` instruction is or how it relates to the loading of the VFT offset as per the above description.

[1] https://github.com/eclipse/openj9/blob/dd96e6747e1a51ca757d501a7c6bf110f108f192/runtime/codert_vm/znathelp.m4#L621-L633
[2] https://github.com/eclipse/openj9/blob/dd96e6747e1a51ca757d501a7c6bf110f108f192/runtime/codert_vm/cnathelp.cpp#L2733-L2752
[3] https://github.com/eclipse/openj9/blob/dd96e6747e1a51ca757d501a7c6bf110f108f192/runtime/oti/JITInterface.hpp#L217-L219
[4] https://github.com/eclipse/openj9/blob/dd96e6747e1a51ca757d501a7c6bf110f108f192/runtime/codert_vm/cnathelp.cpp#L2751
[5] https://github.com/eclipse/openj9/blob/dd96e6747e1a51ca757d501a7c6bf110f108f192/runtime/codert_vm/znathelp.m4#L632
[6] https://github.com/eclipse/openj9/blob/dd96e6747e1a51ca757d501a7c6bf110f108f192/runtime/codert_vm/znathelp.m4#L604-L619
[7] https://github.com/eclipse/openj9/blob/dd96e6747e1a51ca757d501a7c6bf110f108f192/runtime/codert_vm/znathelp.m4#L615-L617

Reply:
I'll look into it more today to get back to were I was when I was first investigating this in regards to that LA instruction comment. But for now should we start testing this code to make sure we are ready to merge assume that all we have to change is some comments?

Reply:
I'm going to merge the PR as is. The comment will be fixed once #8999 is addressed. Thanks!